File: cpp/iteration.cc
#include <iostream>
#include <sys/time.h>
#include <vector>
#include "libpmemkv.h"

#define LOG(msg) std::cout << msg << "\n"

using namespace pmemkv;

static const int COUNT = 100000000;
static const string PATH = "/dev/shm/pmemkv";

static double current_seconds() {
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return (tv.tv_sec) + (double) (tv.tv_usec) / 1000000;
}

struct CallbackContext {
    unsigned long failures;
};

void test_engine(const string engine, const string value) {
    LOG("\nTesting " << engine << " for " + to_string(COUNT) << " keys, value size is "
                     << value.length() << "...");
    std::remove(PATH.c_str());
    KVEngine* kv = KVEngine::Start(engine, "{\"path\":\"" + PATH + "\",\"size\":16106127360}");

    LOG("Put (sequential series)");
    auto started = current_seconds();
    for (int i = 0; i < COUNT; i++) {
        if (kv->Put(to_string(i), value) != OK) {
            std::cout << "Out of space at key " << to_string(i) << "\n";
            exit(-42);
        }
    }
    LOG("   in " << current_seconds() - started << " sec");

    LOG("All (one pass)");
    CallbackContext cxt = {COUNT};
    auto cba = [](void* context, int keybytes, const char* key) {
        ((CallbackContext*) context)->failures--;
    };
    started = current_seconds();
    kv->All(&cxt, cba);
    LOG("   in " << current_seconds() - started << " sec, failures=" + to_string(cxt.failures));

    LOG("Each (one pass)");
    cxt = {COUNT};
    auto cb = [](void* context, int keybytes, const char* key, int valuebytes, const char* value) {
        ((CallbackContext*) context)->failures--;
    };
    started = current_seconds();
    kv->Each(&cxt, cb);
    LOG("   in " << current_seconds() - started << " sec, failures=" + to_string(cxt.failures));

    delete kv;
}

int main() {
    test_engine("tree3", std::string(64, 'A'));
    LOG("\nFinished!\n");
    return 0;
}

File: python/iteration.py
from pmemkv import Database
import os
import time

count = 100000000
failures = count
path = r"/dev/shm/pmemkv"

def func(key = "", value = ""):
    global failures
    failures -= 1

def test_engine(engine, value):
    global count, path, failures

    print ("\nTesting {} engine for {} keys, value size is {}...".format(engine, count, len(value)))

    if os.path.isfile(path):
        os.remove(path)

    db = Database(engine, '{"path": \"' + str(path) + '\", "size": 1073741824, "force_create": 1}')
    print ("Put (sequential series)")
    t1 = time.time()
    for i in range(0, count):
        db.put(str(i), value)
    print ("   in "+ str(time.time() - t1) + " sec(s)")

    print ("All strings (one pass)")
    t1 = time.time()
    db.get_keys_strings(func)
    print ("   in " + str(time.time() - t1) + " sec(s), failures=" + str(failures))
    failures = count

    print ("All (one pass)")
    t1 = time.time()
    db.get_keys(func)
    print ("   in " + str(time.time() - t1) + " sec(s), failures=" + str(failures))
    failures = count

    print ("Each string (one pass)")
    t1 = time.time()
    db.get_all_string(func)
    print ("   in " + str(time.time() - t1) + " sec(s), failures=" + str(failures))
    failures = count

    print ("Each (one pass)")
    t1 = time.time()
    db.get_all(func)
    print ("   in "+  str(time.time() - t1) + " sec(s), failures=" + str(failures))
    failures = count

    db.stop()

test_engine(r"cmap", r"A" * 64)

print ("\nFinished!\n\n")
